File: sdk/python/aep/semver.py
"""
AEP SemVer Matcher — MUST match TypeScript implementation
Reference: spec/core/002-envelope.md §Capability Reference

Supports:
  exact   "1.2.3"
  caret   "^1.2"     >=1.2.0 <2.0.0   (or  >=0.2.0 <0.3.0 for 0.x)
  tilde   "~1.2.3"   >=1.2.3 <1.3.0
  range   ">=1.0.0 <2.0.0"
  star    "*"
  or      "1.2.3 || 1.5.0"
"""

from __future__ import annotations
import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class SemVer:
    major: int
    minor: int
    patch: int
    prerelease: list = None

    def __post_init__(self):
        if self.prerelease is None:
            self.prerelease = []


def normalize_version(version):
    m = re.match(r"^(\d+)(?:\.(\d+))?(?:\.(\d+))?(.*)$", version)
    if not m:
        return version
    major = m.group(1)
    minor = m.group(2) or "0"
    patch = m.group(3) or "0"
    rest = m.group(4) or ""
    return f"{major}.{minor}.{patch}{rest}"


def parse_semver(version):
    normalized = normalize_version(version)
    m = re.match(
        r"^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$",
        normalized,
    )
    if not m:
        return None
    prerelease = m.group(4).split(".") if m.group(4) else []
    return SemVer(
        major=int(m.group(1)),
        minor=int(m.group(2)),
        patch=int(m.group(3)),
        prerelease=prerelease,
    )


def _compare_prerelease(a, b):
    if not a and not b:
        return 0
    if not a:
        return 1
    if not b:
        return -1
    for i in range(max(len(a), len(b))):
        ai = a[i] if i < len(a) else None
        bi = b[i] if i < len(b) else None
        if ai is None:
            return -1
        if bi is None:
            return 1
        an = ai.isdigit()
        bn = bi.isdigit()
        if an and bn:
            diff = int(ai) - int(bi)
            if diff != 0:
                return diff
        elif an:
            return -1
        elif bn:
            return 1
        else:
            if ai < bi:
                return -1
            if ai > bi:
                return 1
    return 0


def compare_semver(a, b):
    if a.major != b.major:
        return -1 if a.major < b.major else 1
    if a.minor != b.minor:
        return -1 if a.minor < b.minor else 1
    if a.patch != b.patch:
        return -1 if a.patch < b.patch else 1
    return _compare_prerelease(a.prerelease, b.prerelease)


@dataclass
class Comparator:
    op: str
    version: SemVer


def _parse_comparator(s):
    m = re.match(r"^(>=|<=|>|<|=)?\s*(\d+(?:\.\d+)?(?:\.\d+)?(?:-[0-9A-Za-z-]+)?(?:\+[0-9A-Za-z-]+)?)$", s)
    if not m:
        return None
    v = parse_semver(m.group(2))
    if v is None:
        return None
    return Comparator(op=m.group(1) or "=", version=v)


def _matches_comparator(v, c):
    cmp = compare_semver(v, c.version)
    if c.op == "=":
        return cmp == 0
    if c.op == ">":
        return cmp > 0
    if c.op == ">=":
        return cmp >= 0
    if c.op == "<":
        return cmp < 0
    if c.op == "<=":
        return cmp <= 0
    return False


def _caret_range(version):
    if version.major > 0:
        return [
            Comparator(">=", version),
            Comparator("<", SemVer(version.major + 1, 0, 0)),
        ]
    if version.minor > 0:
        return [
            Comparator(">=", version),
            Comparator("<", SemVer(0, version.minor + 1, 0)),
        ]
    return [
        Comparator(">=", version),
        Comparator("<", SemVer(0, 0, version.patch + 1)),
    ]


def _tilde_range(version):
    return [
        Comparator(">=", version),
        Comparator("<", SemVer(version.major, version.minor + 1, 0)),
    ]


def _parse_range(range_str):
    or_groups = range_str.split("||")
    result = []
    for part in or_groups:
        trimmed = part.strip()
        if trimmed == "*" or trimmed == "":
            result.append([])
            continue
        if trimmed.startswith("^"):
            v = parse_semver(trimmed[1:].strip())
            if v is None:
                result.append([])
                continue
            result.append(_caret_range(v))
            continue
        if trimmed.startswith("~"):
            v = parse_semver(trimmed[1:].strip())
            if v is None:
                result.append([])
                continue
            result.append(_tilde_range(v))
            continue
        if re.match(r"^\d+\.\d+\.\d+", trimmed):
            v = parse_semver(trimmed)
            if v is None:
                result.append([])
                continue
            result.append([Comparator("=", v)])
            continue
        comps = []
        for token in trimmed.split():
            c = _parse_comparator(token)
            if c is not None:
                comps.append(c)
        result.append(comps)
    return result


def satisfies(version, range_str):
    v = parse_semver(version)
    if v is None:
        return False
    if range_str == "*" or range_str == "":
        return True
    or_groups = _parse_range(range_str)
    return any(
        len(group) == 0 or all(_matches_comparator(v, c) for c in group)
        for group in or_groups
    )

File: sdk/rust/src/semver.rs
//! AEP SemVer Matcher — MUST match TypeScript and Python implementations.
//!
//! Reference: spec/core/002-envelope.md §Capability Reference
//!
//! Supports:
//!   exact   "1.2.3"
//!   caret   "^1.2"     >=1.2.0 <2.0.0   (or  >=0.2.0 <0.3.0 for 0.x)
//!   tilde   "~1.2.3"   >=1.2.3 <1.3.0
//!   range   ">=1.0.0 <2.0.0"
//!   star    "*"
//!   or      "1.2.3 || 1.5.0"

use regex::Regex;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SemVer {
    pub major: u32,
    pub minor: u32,
    pub patch: u32,
    pub prerelease: Vec<String>,
}

/// Normalize incomplete version strings:
///   "1"     → "1.0.0"
///   "1.2"   → "1.2.0"
///   "1.2.3" → "1.2.3"
fn normalize_version(version: &str) -> String {
    let re = Regex::new(r"^(\d+)(?:\.(\d+))?(?:\.(\d+))?(.*)$").unwrap();
    if let Some(caps) = re.captures(version) {
        let major = caps.get(1).map(|m| m.as_str()).unwrap_or("0");
        let minor = caps.get(2).map(|m| m.as_str()).unwrap_or("0");
        let patch = caps.get(3).map(|m| m.as_str()).unwrap_or("0");
        let rest = caps.get(4).map(|m| m.as_str()).unwrap_or("");
        return format!("{}.{}.{}{}", major, minor, patch, rest);
    }
    version.to_string()
}

/// Parse a version string into SemVer. Returns None on failure.
pub fn parse_semver(version: &str) -> Option<SemVer> {
    let normalized = normalize_version(version);
    let re = Regex::new(
        r"^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$",
    )
    .unwrap();
    let caps = re.captures(&normalized)?;
    let prerelease = caps
        .get(4)
        .map(|m| m.as_str().split('.').map(|s| s.to_string()).collect())
        .unwrap_or_default();
    Some(SemVer {
        major: caps.get(1)?.as_str().parse().ok()?,
        minor: caps.get(2)?.as_str().parse().ok()?,
        patch: caps.get(3)?.as_str().parse().ok()?,
        prerelease,
    })
}

fn compare_prerelease(a: &[String], b: &[String]) -> std::cmp::Ordering {
    if a.is_empty() && b.is_empty() {
        return std::cmp::Ordering::Equal;
    }
    if a.is_empty() {
        return std::cmp::Ordering::Greater; // no prerelease > has prerelease
    }
    if b.is_empty() {
        return std::cmp::Ordering::Less;
    }
    let max_len = a.len().max(b.len());
    for i in 0..max_len {
        let ai = a.get(i);
        let bi = b.get(i);
        match (ai, bi) {
            (None, Some(_)) => return std::cmp::Ordering::Less,
            (Some(_), None) => return std::cmp::Ordering::Greater,
            (None, None) => return std::cmp::Ordering::Equal,
            (Some(a_str), Some(b_str)) => {
                let a_num = a_str.parse::<i64>().ok();
                let b_num = b_str.parse::<i64>().ok();
                match (a_num, b_num) {
                    (Some(an), Some(bn)) => {
                        if an != bn {
                            return an.cmp(&bn);
                        }
                    }
                    (Some(_), None) => return std::cmp::Ordering::Less, // numeric < non-numeric
                    (None, Some(_)) => return std::cmp::Ordering::Greater,
                    (None, None) => {
                        match a_str.cmp(b_str) {
                            std::cmp::Ordering::Equal => continue,
                            other => return other,
                        }
                    }
                }
            }
        }
    }
    std::cmp::Ordering::Equal
}

pub fn compare_semver(a: &SemVer, b: &SemVer) -> std::cmp::Ordering {
    if a.major != b.major {
        return a.major.cmp(&b.major);
    }
    if a.minor != b.minor {
        return a.minor.cmp(&b.minor);
    }
    if a.patch != b.patch {
        return a.patch.cmp(&b.patch);
    }
    compare_prerelease(&a.prerelease, &b.prerelease)
}

#[derive(Debug, Clone)]
struct Comparator {
    op: String, // ">=", "<=", ">", "<", "="
    version: SemVer,
}

fn matches_comparator(v: &SemVer, c: &Comparator) -> bool {
    let cmp = compare_semver(v, &c.version);
    match c.op.as_str() {
        "=" => cmp == std::cmp::Ordering::Equal,
        ">" => cmp == std::cmp::Ordering::Greater,
        ">=" => matches!(cmp, std::cmp::Ordering::Greater | std::cmp::Ordering::Equal),
        "<" => cmp == std::cmp::Ordering::Less,
        "<=" => matches!(cmp, std::cmp::Ordering::Less | std::cmp::Ordering::Equal),
        _ => false,
    }
}

fn parse_comparator(s: &str) -> Option<Comparator> {
    let re = Regex::new(r"^(>=|<=|>|<|=)?\s*(\d+(?:\.\d+)?(?:\.\d+)?(?:-[0-9A-Za-z-]+)?(?:\+[0-9A-Za-z-]+)?)$").unwrap();
    let caps = re.captures(s)?;
    let v = parse_semver(caps.get(2)?.as_str())?;
    let op = caps.get(1).map(|m| m.as_str().to_string()).unwrap_or_else(|| "=".to_string());
    Some(Comparator { op, version: v })
}

fn caret_range(version: &SemVer) -> Vec<Comparator> {
    if version.major > 0 {
        vec![
            Comparator { op: ">=".to_string(), version: version.clone() },
            Comparator { op: "<".to_string(), version: SemVer { major: version.major + 1, minor: 0, patch: 0, prerelease: vec![] } },
        ]
    } else if version.minor > 0 {
        vec![
            Comparator { op: ">=".to_string(), version: version.clone() },
            Comparator { op: "<".to_string(), version: SemVer { major: 0, minor: version.minor + 1, patch: 0, prerelease: vec![] } },
        ]
    } else {
        vec![
            Comparator { op: ">=".to_string(), version: version.clone() },
            Comparator { op: "<".to_string(), version: SemVer { major: 0, minor: 0, patch: version.patch + 1, prerelease: vec![] } },
        ]
    }
}

fn tilde_range(version: &SemVer) -> Vec<Comparator> {
    vec![
        Comparator { op: ">=".to_string(), version: version.clone() },
        Comparator { op: "<".to_string(), version: SemVer { major: version.major, minor: version.minor + 1, patch: 0, prerelease: vec![] } },
    ]
}

fn parse_range(range_str: &str) -> Vec<Vec<Comparator>> {
    range_str
        .split("||")
        .filter_map(|part| {
            let trimmed = part.trim();
            if trimmed.is_empty() || trimmed == "*" {
                return Some(vec![]);
            }
            if trimmed.starts_with('^') {
                let v = parse_semver(trimmed[1..].trim())?;
                return Some(caret_range(&v));
            }
            if trimmed.starts_with('~') {
                let v = parse_semver(trimmed[1..].trim())?;
                return Some(tilde_range(&v));
            }
            // bare version = exact
            let bare_re = Regex::new(r"^\d+\.\d+\.\d+").unwrap();
            if bare_re.is_match(trimmed) {
                let v = parse_semver(trimmed)?;
                return Some(vec![Comparator { op: "=".to_string(), version: v }]);
            }
            // space-separated comparators
            let comps: Vec<Comparator> = trimmed
                .split_whitespace()
                .filter_map(|t| parse_comparator(t))
                .collect();
            Some(comps)
        })
        .collect()
}

/// Check if version satisfies range. Matches TS satisfies() and Python satisfies().
pub fn satisfies(version: &str, range_str: &str) -> bool {
    let v = match parse_semver(version) {
        Some(v) => v,
        None => return false,
    };
    if range_str == "*" || range_str.is_empty() {
        return true;
    }
    let or_groups = parse_range(range_str);
    or_groups.iter().any(|group| {
        group.is_empty() || group.iter().all(|c| matches_comparator(&v, c))
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exact_match() {
        assert!(satisfies("1.2.3", "1.2.3"));
        assert!(!satisfies("1.2.4", "1.2.3"));
    }

    #[test]
    fn test_caret_in_range() {
        assert!(satisfies("1.5.0", "^1.2"));
        assert!(satisfies("1.2.0", "^1.2"));
        assert!(!satisfies("2.0.0", "^1.2"));
    }

    #[test]
    fn test_caret_0_x_minor() {
        assert!(satisfies("0.2.5", "^0.2.3"));
        assert!(!satisfies("0.3.0", "^0.2.3"));
    }

    #[test]
    fn test_caret_0_0_x_patch() {
        assert!(satisfies("0.0.3", "^0.0.3"));
        assert!(!satisfies("0.0.4", "^0.0.3"));
    }

    #[test]
    fn test_tilde_in_range() {
        assert!(satisfies("1.2.5", "~1.2.3"));
        assert!(!satisfies("1.3.0", "~1.2.3"));
    }

    #[test]
    fn test_star() {
        assert!(satisfies("99.99.99", "*"));
    }

    #[test]
    fn test_or() {
        assert!(satisfies("1.2.3", "1.2.3 || 1.5.0"));
        assert!(satisfies("1.5.0", "1.2.3 || 1.5.0"));
        assert!(!satisfies("1.4.0", "1.2.3 || 1.5.0"));
    }

    #[test]
    fn test_range() {
        assert!(satisfies("1.5.0", ">=1.0.0 <2.0.0"));
        assert!(!satisfies("2.0.0", ">=1.0.0 <2.0.0"));
        assert!(!satisfies("0.9.0", ">=1.0.0 <2.0.0"));
    }

    #[test]
    fn test_incomplete_major() {
        assert!(satisfies("1.0.0", "1"));
    }

    #[test]
    fn test_incomplete_minor() {
        assert!(satisfies("1.2.0", "1.2"));
    }
}
